### PR TITLE
Winter adaptation for Ectothermic

### PR DESCRIPTION
**Pull request by Kamejeir**
#### Summary
Features "Cold-blooded winter adaptation"

#### Purpose of change

[Anid proposed](https://github.com/CleverRaven/Cataclysm-DDA/issues/66539#issuecomment-1614076189) in #66539 to make cold-blooded creatures enter a sort of homeostasis with what Fairyarmadillo mentioned, which was never followed up on as far as I can see.

#### Describe the solution

Make Ectothermic not take frostbite damage, at the cost of having their uptime in the winter decrease by making them a lot more sleepy.

Redid the mutation path of Temperature Tolerance, including Ectothermic as part of it, as a part of the issue while also fixing flip-flopping (between Heat Tolerance and Temperature Tolerance).

//

~~While checking through cold-blooded creatures for this, I saw and fixed bark blister bonuses not being applied for upgrades (together with preventing the non-purifiable mutation from seemingly downgrading in very esoteric circumstances involving getting a skin mutation).~~

~~Also only slightly on the subject of cold-blooded creatures, Muscle Consolidation was weirdly "missable" by the random order of getting Gastropod's mutations dictating the outcome you get.~~

#### Describe alternatives you've considered

I could make Ectothermic mutants straight-up COLD_IMMUNE and set up every single missed check in respect to that instead, but I don't know what kind of balance issues that might cause if some future Erk dungeon enemy saps your body heat to harm you...

#### Testing

This coming out of draft will imply having tested all facets of getting, purifying and seeing that the mutations involved work properly.

#### Additional context

This does cause a total of -5 chilling with Temperature Tolerance, due to Heat Tolerance being made to disappear in the process. If this counts as a regression, I'll amend that by increasing the relevant number.

The speed modifier could and should drastically increase as to compensate the "chilly" effect not being present.

### CODE REVIEW

**Comment by Fris0uman** [comment]
Ectotherms being resistant to cold feels like the exact opposite of what ectotherm means. It's weird that the inability to regulate your body temperature gives you resistance to cold. Some amphibian are able to freeze and remain alive but it's not a generale rule, and I think that other ectotherms either die off, and only the eggs of the next generation go through winter, or bury themselves to stay at low but constant and above freezing temperature.

File: data/json/mutations/mutations.json [hunk 1]
```diff
@@ -3424,7 +3424,7 @@
     "ugliness": 3,
     "description": "All the hair on your body has turned to long, grass-like leaves.  Apart from being physically striking, these provide you with a minor amount of nutrients while in sunlight when your head is uncovered.  Slightly reduces wet effects.",
     "types": [ "HAIR" ],
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:3427]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 2]
```diff
@@ -3470,7 +3470,7 @@
     "visibility": 0,
     "ugliness": 0,
     "description": "You body has begun moving nutrients via the evaporation of water.  This increases your thirst when it's hot, but reduces it when it's cold.",
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:3473]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 3]
```diff
@@ -3483,7 +3483,7 @@
     "visibility": 10,
     "ugliness": -4,
     "description": "You've started blooming where your scalp used to be.  Your blossoms have a pleasant aroma, are visually striking, and are quite sensitive.",
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:3486]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 4]
```diff
@@ -4035,7 +4035,7 @@
     "points": 1,
     "visibility": 1,
     "description": "Your extremities have gained strength but lost fine motor skills.  -1 Dexterity +2 Strength.",
-    "changes_to": [ "GASTROPOD_EXTREMITY2" ],
+    "leads_to": [ "GASTROPOD_EXTREMITY2" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:4038]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 5]
```diff
@@ -7497,7 +7497,7 @@
     "ugliness": 5,
     "description": "You have developed several vines sprouting from your shoulder area.  They're bulky, and get in the way.",
     "types": [ "WINGS" ],
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:7500]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 6]
```diff
@@ -7511,7 +7511,7 @@
     "ugliness": 5,
     "description": "You've developed the ability to control your vines; they make good lashes.  You can even rappel down sheer drops using them, but disconnecting HURTS.",
     "types": [ "WINGS" ],
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:7514]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 7]
```diff
@@ -7534,7 +7534,7 @@
     "ugliness": 6,
     "description": "You have full control of your vines, and can grow new ones and detach old ones more or less at will.",
     "types": [ "WINGS" ],
-    "prereqs": [ "PLANTSKIN", "BARK" ],
+    "prereqs": [ "PLANTSKIN", "BARK", "BARK2_a", "BARK2_b", "BARK2_c" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:7537]: This seem unrelated to the ectotherm change and should go in a different PR

File: data/json/mutations/mutations.json [hunk 8]
```diff
@@ -9533,6 +9535,7 @@
     "description": "You are now even less bothered by the heat and it is harder for you to overheat or even sweat.  When you do though, it'll hit just as hard as it did before.",
     "prereqs": [ "RESISTWARM" ],
     "types": [ "ACCLIMATIZATION" ],
+    "changes_to": [ "RESISTCHILL_RESISTWARM" ],
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:9538]: How odes this relate to the change to ECTOTHERM?

File: data/json/mutations/mutations.json [hunk 9]
```diff
@@ -9587,7 +9590,9 @@
     "points": 5,
     "valid": true,
     "description": "You are now not at all bothered by the heat and it is harder for you to overheat or freeze at all.  When you do though, it'll hit just as hard as it did before.",
-    "prereqs": [ "RESISTWARM" ],
+    "prereqs": [ "STRONGER_RESISTWARM" ],
+    "prereqs2": [ "COLDBLOOD4" ],
+    "purifiable": false,
```
Comment (Fris0uman) [inline data/json/mutations/mutations.json:9595]: Same question, how does this relate to the goal of this PR?

File: src/character_body.cpp [hunk 10]
```diff
@@ -102,6 +102,9 @@ static const json_character_flag json_flag_NO_THIRST( "NO_THIRST" );
 static const json_character_flag json_flag_PAIN_IMMUNE( "PAIN_IMMUNE" );
 
 static const trait_id trait_BARK( "BARK" );
+static const trait_id trait_BARK2_a( "BARK2_a" );
+static const trait_id trait_BARK2_b( "BARK2_b" );
+static const trait_id trait_BARK2_c( "BARK2_c" );
```
Comment (Fris0uman) [inline src/character_body.cpp:107]: unrelated and should go in different PR

File: src/character_body.cpp [hunk 11]
```diff
@@ -458,7 +461,7 @@ void Character::update_bodytemp()
     int bp_windpower = get_local_windpower( weather_man.windspeed + vehwindspeed, cur_om_ter,
                                             get_location(), weather_man.winddirection, sheltered );
     // Let's cache this not to check it for every bodyparts
-    const bool has_bark = has_trait( trait_BARK );
+    const bool has_bark = has_trait( trait_BARK ) || has_trait( trait_BARK2_a ) || has_trait( trait_BARK2_b ) || has_trait( trait_BARK2_c );
```
Comment (Fris0uman) [inline src/character_body.cpp:464]: unrelated and should go in different PR

**Comment by fairyarmadillo** [comment]
Instead of making them immune, simply granting a degree of frostbite resistance via cold tolerant would probably allow gastropods, lizards, and amphibians the ability to sleep in cold temperatures without making them immune to freezing entirely.

https://www.npr.org/2024/01/25/1226660910/alligators-frozen-underwater-brumation-explanation

https://www.molluscs.at/gastropoda/index.html?/gastropoda/dormancy.html

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4331536/

https://www.canr.msu.edu/news/how_insects_survive_cold_the_potential_effect_of_a_mild_winter

These abilities are very widespread among ectotherms. They are not universal, but there are many ways these could be implemented.

**Comment by Kamejeir** [comment]
@fairyarmadillo I'm trying very hard to listen to what you're saying. My best guess is that you suggest Ectothermic to be a cold-blooded Hibernation equivalent that grants temporary frostbite immunity.

**Comment by fairyarmadillo** [comment]
> @fairyarmadillo I'm trying very hard to listen to what you're saying. My best guess is that you suggest Ectothermic to be a cold-blooded Hibernation equivalent that grants temporary frostbite immunity.

My original plan was to make a mutation that would protect ectothermic characters only during their dormancy, assuming they met all the necessary conditions, but that would probably require some c++.

**Comment by Kamejeir** [comment]
Granular protection is harder than the immune flag, but I'll see what I can do to honor your wishes.

**Comment by fairyarmadillo** [comment]
> to honor your wishes.

Don't worry too much about that, they were just ideas I had! :)

**Comment by Kamejeir** [comment]
So, "Dormancy".
I was beginning to code checking for applying frostbite resistance during the times they were sleeping, but then it struck me that cold-blooded creatures might instead become totally inactive during winter.

That is outside of my original scope, and I don't know what to do, as it was never my intention for them to essentially skip winter (and all the cold problems that come with it).

**Comment by fairyarmadillo** [comment]
> That is outside of my original scope, and I don't know what to do, as it was never my intention for them to essentially skip winter (and all the cold problems that come with it).

That _was_ my intention. You'd fix up a safe space and ingest a huge amount of calories before passing out for several weeks. This would run the risk of getting you killed if an enemy found you while you were unconscious (if you woke early you'd have a big debuff) and in any event you'd awaken at the end of it in a massively malnourished state.

You wouldn't skip all of winter, but you'd be able to skip the worst bit of it. This might be doable with EOCs and enchantments alone nowadays, but I'm not sure.

It would be huge project in any event. And I'm only sharing this so you know what my original intent was, there's nothing stopping you from trying something else.

**Comment by Kamejeir** [comment]
Well, thank you for your input. It has been cool discussing this with you, however, it seems both Fris and Erk are opposing the original idea. As you say, it has turned into a "huge project" to get this anywhere I wanted... I'm sorry, but I think I'll give up.
